File: archive/e/erlang/bubble_sort.erl
-module(bubble_sort).
-export([main/1]).

%% Run with: escript bubble_sort.erl <INPUT>

usage() ->
    io:format("Usage: please provide a list of at least two integers to sort in the format \"1, 2, 3, 4, 5\"~n"),
    halt().

convert_list_to_integers(Str) ->
    Values = string:tokens(Str, ","),
    NumCommas = count_commas(Str),
    NumValues = length(Values),
    if
        (NumCommas + 1) == NumValues ->
            convert_values(Values);
        true ->
            {error, []}
    end.

count_commas([]) ->
    0;
count_commas([$, | Rest]) ->
    1 + count_commas(Rest);
count_commas([_ | Rest]) ->
    count_commas(Rest).

convert_values([]) ->
    {ok, []};
convert_values([ValueStr | Rest]) ->
    case convert_to_integer(ValueStr) of
        {ok, Value} ->
            case convert_values(Rest) of
                {ok, RestValues} ->
                    {ok, [Value | RestValues]};
                _ ->
                    {error, []}
            end;
        _ ->
            {error, []}
    end.

convert_to_integer(Str) ->
    Result = catch string:to_integer(string:strip(Str)),
    case Result of
        {Int, Rest} when Rest == "" ->
            {ok, Int};
        _ ->
            {error, 0}
    end.

bubble_sort(List) -> bubble_sort(List, [], false).

bubble_sort([A, B | T], Acc, _) when A > B ->
    bubble_sort([A | T], [B | Acc], true);
bubble_sort([A, B | T], Acc, Tainted) ->
    bubble_sort([B | T], [A | Acc], Tainted);
bubble_sort([A | T], Acc, Tainted) ->
    bubble_sort(T, [A | Acc], Tainted);
bubble_sort([], Acc, true) ->
    bubble_sort(lists:reverse(Acc));
bubble_sort([], Acc, false) ->
    lists:reverse(Acc).


handle_output([Num]) ->
    io:format("~p~n", [Num]);
handle_output([Num|NumList]) ->
    io:format("~p, ", [Num]),
    handle_output(NumList).

main(Args) ->
    if
        length(Args) < 1 ->
            usage();
        true ->
            ok
    end,

    StrValue = lists:nth(1, Args),
    Values = case convert_list_to_integers(StrValue) of
        {ok, Result} ->
            Result;
        _ ->
            usage()
    end,

    if
        length(Values) < 2 ->
            usage();
        true ->
            ok
    end,

    SortedValues = bubble_sort(Values),
    handle_output(SortedValues).

File: archive/e/elixir/bubble-sort.ex
defmodule Bubble do
  @doc """
  Sort the giveen input using bubble sort
  """
  def main(args) do
    cond do
      length(args) != 1 
      or List.first(args) == "" 
      or List.first(args) |> String.split(",") |> length < 2 ->
        IO.puts("Usage: please provide a list of at least two integers to sort in the format \"1, 2, 3, 4, 5\"")
      true ->
        input = List.first(args)
        |> String.split(",") 
        |> Enum.map(&String.trim/1)
        |> Enum.map(&String.to_integer/1)
        sorted_list = pass(input, [])
        IO.puts(Enum.join(sorted_list, ", "))
    end
  end

  @doc """
  Handle each pass, return the list when there are no more changes to the list
  """
  def pass(new_list, old_list) when new_list == old_list, do: new_list

  def pass(new_list, old_list) when new_list != old_list do
    passed_list = sort(new_list, [])
    pass(passed_list, new_list)
  end

  @doc """
  Add the smaller element to accumulator, recursively call sort until all elements are added to accumulator
  """
  def sort([], acc), do: acc

  def sort([head, tail], acc) do
    if head > tail, do: sort([], acc ++ [tail, head]), else: sort([], acc ++ [head, tail])
  end

  def sort([head | [head2 | tail]], acc) do
    if head > head2, do: sort([head | tail], acc ++ [head2]), else: sort([head2 | tail], acc ++ [head])
  end
end

Bubble.main(System.argv())
